crates/contracts/src/model/definition.cairo

```diff
@@ -0,0 +1,20 @@
+use dojo::meta::Layout;
+
+#[derive(Copy, Drop, Serde, Debug, PartialEq)]
+pub enum ModelIndex {
+    Keys: Span<felt252>,
+    Id: felt252,
+    // (entity_id, member_id)
+    MemberId: (felt252, felt252)
+}
+
+pub trait ModelDefinition<T> {
```

Comment:
@glihm @remybar the newest update to main now has a Struct called ModelDefinition which will cause conflicts, What do you think the best way to resolve this is? Is the ModelDefinition struct even necessary/ever used?

Reply:
Hi @bengineer42 !
It is the same trait than the one you added in your PR but coming from this merged PR (https://github.com/dojoengine/dojo-core/pull/10).
I think there is no conflict, it should be the same trait 🤔

Reply:
Isnt the `ModelDefinition` in that PR a struct not a trait?

Reply:
> Isnt the `ModelDefinition` in that PR a struct not a trait?

Ah yes, understood 👍 
Yes, this `ModelDefinition` struct is used to return the full model definition from the model contract, so you can get the whole definition with only one contract call instead of calling all the functions one by one.

Maybe we can juste rename the `struct ModelDefinition` to something else ?

Reply:
Any suggestions? ModelAttributes, @glihm I know you didn't like this for the trait but  I'm lost for ideas

Reply:
> Any suggestions? ModelAttributes, @glihm I know you didn't like this for the trait but I'm lost for ideas

`ModelDef` ? 😅 
In fact, it provides the same information than the `ModelDefinition` trait ... So quite difficult to find a different name for that.

Reply:
I think it is better to keep `ModelDefinition` for the trait as it will be used directly, while the struct name won't be used. We will do something like:
```rust
    let definition = dispatcher.definition();
    println!("{}", definition.name);
```

Reply:
I will change ti to ModelDef for now just to keep it moving but if anyone has any better suggestions let me know

Reply:
@remybar there are some inconsistencies between struct and trait, eg  `fn namespace_hash() -> felt252;` vs `pub namespace_selector: felt252,`

```rust
pub trait ModelDefinition<T> {
    fn name() -> ByteArray;
    fn namespace() -> ByteArray;
    fn tag() -> ByteArray;
    fn version() -> u8;
    fn selector() -> felt252;
    fn name_hash() -> felt252;
    fn namespace_hash() -> felt252;
    fn layout() -> Layout;
}


#[derive(Drop, Serde, Debug, PartialEq)]
pub struct ModelDef {
    pub name: ByteArray,
    pub namespace: ByteArray,
    pub namespace_selector: felt252,
    pub version: u8,
    pub layout: Layout,
    pub schema: Ty,
    pub packed_size: Option<u32>,
    pub unpacked_size: Option<u32>
}
```

Reply:
> @remybar there are some inconsistencies between struct and trait, eg `fn namespace_hash() -> felt252;` vs `pub namespace_selector: felt252,`
> 
> ```rust
> pub trait ModelDefinition<T> {
>     fn name() -> ByteArray;
>     fn namespace() -> ByteArray;
>     fn tag() -> ByteArray;
>     fn version() -> u8;
>     fn selector() -> felt252;
>     fn name_hash() -> felt252;
>     fn namespace_hash() -> felt252;
>     fn layout() -> Layout;
> }
> 
> 
> #[derive(Drop, Serde, Debug, PartialEq)]
> pub struct ModelDef {
>     pub name: ByteArray,
>     pub namespace: ByteArray,
>     pub namespace_selector: felt252,
>     pub version: u8,
>     pub layout: Layout,
>     pub schema: Ty,
>     pub packed_size: Option<u32>,
>     pub unpacked_size: Option<u32>
> }
> ```

You can change to `namespace_hash` in the struct, thanks 👍 Maybe we should also add `name_hash` as it exists in the `ModelDefinition` trait.